### DISCUSSION TITLE
Universal blocks and block hydration

### DISCUSSION THREAD

**Discussion by michalczaplinski** [WordPress/gutenberg#38224] [Developer Experience]
> **TL;DR**
>
> When you create a dynamic block, there is duplication of logic between PHP and JS code. I propose that we should try to achieve "universal blocks" that could be written once and run in both PHP and JS with the help of a compiler. Additionally, I explore how this idea intersects with block hydration, which is a way to add interactivity to server-rendered blocks. Finally, I show a basic <a href="https://github.com/michalczaplinski/hydration-mitosis-demo">demo</a> of compiling blocks authored in <a href="https://github.com/BuilderIO/mitosis">Mitosis</a> format to both <a href="https://shopify.github.io/liquid/">Liquid templates</a> <em>and</em> React components that can hydrate the server-rendered markup.
> 

<h2 id="challenges">The challenges</h2>

<h3 id="1-duplicate-logic-in-php-and-js-when-writing-dynamic-blocks-universal-blocks">1. Duplicate logic in PHP and JS when writing dynamic blocks (universal blocks)</h3>

<p>When writing <em>dynamic blocks</em>, there is often necessary <strong>duplication of logic</strong> between the <code><a href="https://developer.wordpress.org/block-editor/reference-guides/block-api/block-edit-save/#edit">edit</a></code> function of the block (written in javascript) and the function called by the <a href="https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/creating-dynamic-blocks/">render_callback</a> on the server side, typically located in the <code>index.php</code> file of the block. <a href="https://github.com/michalczaplinski/gutenberg/blob/4582d021b997cfba3f270f95f9066b2173c490a5/packages/block-library/src/post-date"></a><a href="https://github.com/Wordpress/gutenberg/tree/4582d021b997cfba3f270f95f9066b2173c490a5/packages/block-library/src/post-date">Post Date</a> <a href="https://github.com/michalczaplinski/gutenberg/blob/4582d021b997cfba3f270f95f9066b2173c490a5/packages/block-library/src/post-date"></a>core block is a simple example of a block with such duplication.</p>

<p>Unsurprisingly, everyone agrees that this duplication is <strong>bad.</strong> The users would like to see <em>some</em> API which would allow them to write a <strong>single</strong> <strong>template</strong> that could be rendered both on the client and the server. I'll refer to this idea as "universal blocks" throughout the post 🙂</p>

<p>There has already been a fair bit of discussion about this problem in many other places:</p>

<ul id="block-fc3c6f67-a1ca-421c-a597-89808aaf415c"><li><a href="https://github.com/WordPress/gutenberg/issues/14446">Render dynamic blocks on the front end in JavaScript instead of PHP</a> - the main github issue on the topic of "universal blocks". Various people suggest various strategies: Rendering on client only, using a templating language like Handlebars, using webcomponents...</li><li><a href="https://github.com/youknowriad/blocky"></a><a href="https://github.com/youknowriad/blocky">https://github.com/youknowriad/blocky</a> - An experiment by @youknowriad where he explores the concept of "universal blocks" - very interesting. It allows you to create a block using a single JSON file where you define your "template" as well as any attributes and properties, like in a <code>block.json</code> file. AFAIK, this is also where he coined the term "universal blocks"! 😃</li><li><a href="https://github.com/mcsf/prototype-block-generator">https://github.com/mcsf/prototype-block-generator</a> - An experiment by @miguelcsf to generate block HTML from a preact component tree and a template.</li><li><a href="https://twitter.com/markjaquith/status/1432461667187765251"></a><a href="https://twitter.com/markjaquith/status/1432461667187765251">https://twitter.com/markjaquith/status/1432461667187765251</a> - A thread with a discussion of making block creation easier.</li><li><a href="https://wptavern.com/a-world-where-some-block-development-is-merely-a-templating-system-with-no-build-process"></a><a href="https://wptavern.com/a-world-where-some-block-development-is-merely-a-templating-system-with-no-build-process">https://wptavern.com/a-world-where-some-block-development-is-merely-a-templating-system-with-no-build-process</a> - An article about the above twitter thread</li>

<li><a href="https://github.com/markjaquith/BlockStub"></a><a href="https://github.com/markjaquith/BlockStub">https://github.com/markjaquith/BlockStub</a> - An experiment in writing blocks with PHP which would dynamically generate React code.</li>
<li><a href="https://www.youtube.com/watch?v=-PZwfvtkQOM">Live Q &amp; A: How can we make building custom blocks easier?</a> - a video of a panel discussion on the topic of "universal" blocks with Birgit, Helen, Mark and Riad.</li></ul>

<h3 id="2-make-it-easier-to-create-interactive-blocks-block-hydration">2. Make it easier to create interactive blocks (block hydration)</h3>

<p>Users would like to create blocks that use JS to create interactive experiences.</p>

<p>Woocommerce has been able to work with existing constraints to <a href="https://developer.woocommerce.com/2021/11/15/how-does-woocommerce-blocks-render-interactive-blocks-in-the-frontend">create interactive</a> blocks that are being rendered only on the client using React, lazy-loading and some clever techniques. This is a great solution, but we should find a way to support their use case with an even better out-of-the-box experience.</p>

<p>Finding a solution to this challenge should involve figuring out how to <a href="https://en.wikipedia.org/wiki/Hydration_(web_development)">hydrate</a> the blocks generated in PHP. Why? For many blocks, only rendering on the client is not a good solution because it makes for a less-than-ideal UX (you have to wait for all client-side JS to load in order to render your block). So, ideally, we'd like to render the content <em>on the server</em> and then <strong>hydrate</strong> it a.k.a. "make interactive" using JS on the client. How? Well... that's the challenge 🙂.</p>

<h2 id="define-dynamic-blocks-using-a-high-level-block-format-and-compile-them-to-a-format-that-gutenberg-understands">Define dynamic blocks using a high-level block format and compile them to a format that Gutenberg understands</h2>

<p>This is a solution proposed in <a href="https://github.com/youknowriad/blocky">blocky</a>. I think this idea is worth exploring more. To recap: Riad proposed using plain JSON to define blocks, but we could push this idea even further and use a more sophisticated block format and a compiler!</p>

<p>But first, we gotta ask ourselves: "What are our constraints"?</p>

<ul><li>The system that we design should be completely opt-in. That is, the current Gutenberg API and ways to create blocks have to continue to work. This means that the system should work "on top of" the current Gutenberg API.</li><li>We do not have a JS runtime, i.e. <code>node</code> on the server. We can only assume that the block creator has node installed and can use it only at build time (when authoring blocks).</li><li>We cannot use custom PHP extensions beyond ones listed in <a href="https://make.wordpress.org/hosting/handbook/server-environment/#php-extensions">https://make.wordpress.org/hosting/handbook/server-environment/#php-extensions</a>. This means that running <a href="https://github.com/phpv8/v8js">V8JS</a> to server-render react components is out of the question.</li><li><strong>Bonus:</strong> It should be possible to re-use blocks outside of Gutenberg, e.g in Drupal. This use case has been mentioned by Riad <a href="https://www.youtube.com/watch?v=-PZwfvtkQOM">in the Q&amp;A</a> and it ties into the discussion of "universal blocks" he explored in blocky. If we can make blocks in a Gutenberg-agnostic way, they could potentially be used in other contexts.</li><li><strong>Bonus:</strong> Our blocks should work with any framework on the frontend. This means that even though we create blocks with React, they <em>don't have to</em> be React components when they are used on the frontend. Maybe someone is using Vue or Preact in their WordPress site and would like their block to work with the rest of their site. Maybe the blocks could even be Web Component that could be used with any framework (but I'm getting ahead of myself here 😅).</li></ul>

<p>Given those constraints, I've also briefly considered leveraging WebAssembly to compile the <a href="https://bellard.org/quickjs/">QuickJS</a> javascript engine to WASM and run it using <a href="https://github.com/wasmerio/wasmer-php">wasmer-php</a> runtime. However, this turns out to also require a custom PHP extension so is a non-starter at the moment.</p>

<h2 >The demo</h2>

https://github.com/michalczaplinski/hydration-mitosis-demo

<p>I've created a demo <a href="https://github.com/michalczaplinski/hydration-mitosis-demo">here</a> which allows you to create Gutenberg blocks using <a href="https://github.com/BuilderIO/mitosis">Mitosis</a> format and compile them to both a Liquid template AND a React component that hydrates the server-rendered HTML.</p>

<p>There is more information in the repo, including what I'd consider are limitations and weaknesses of this approach 🙂</p>

<h2 id="other-ideas">Other ideas</h2>

<p>There have been other ideas floated around with respect to "universal" blocks so I'd like to give them credit here as well.</p>

<h3 id="render-blocks-on-the-client-only">Render blocks on the client only</h3>

<p>I think this idea is probably a non-starter for SEO, performance and UX reasons, but I include it here for completeness as it does technically solve some challenges. It was proposed <a href="https://github.com/WordPress/gutenberg/issues/14446#issuecomment-527204617">here</a>. </p>

<h3 id="define-blocks-in-php">Define blocks in PHP</h3>

<p>This is the approach proposed by @markjaquith in <a href="https://github.com/markjaquith/BlockStub">https://github.com/markjaquith/BlockStub</a>. I think this is an interesting idea, especially as it might be more approachable to PHP developers over more JS-focused approaches, but I believe it might be hard to scale beyond a "hello world" - I feel that writing JS <em>inside</em> of PHP might be the worst of both worlds though... You still have to write some JS for the interactive parts of your app but you forgo all the DX goodness that come with modern JS dev environment and just concatenate strings in PHP. On the plus side though, this approach would require no build step.</p>

<h2 id="conclusion-and-open-question">Conclusion and open questions</h2>

<p>We are still very early with ideas for "universal blocks" and block hydration so I would 💕 some feedback here and any ideas that might have occurred to anyone. I also want to stress that <strong>the idea to use Mitosis is only an experiment</strong> and it is meant to serve as an example of the <em>kind</em> of direction that we could take here and <strong> NOT a suggestion to adopt it inside Gutenberg.</strong> 

Those are some of the questions that I still have lingering in my mind:</p>

<ul><li>I wonder if there is a place for web components to play a role here, like originally <a href="https://github.com/WordPress/gutenberg/issues/14446#issuecomment-480696030">suggested by</a> @felixarntz . It would be great if we could compile our new block format to web components. This way, we could have framework-agnostic and browser-native block elements. The challenge here I think is finding a way to server-render web components in PHP - not sure if possible, but would love see this investigated more!</li><li><span style="color:initial;font-family:var(--p2-font-body);font-size:1rem;">How could we extend this work to enable creating blocks that mutate some state on the server? For example like a "checkbox" block or a "Comment Form" block? </span> An approach similar to React Server Components is one idea. I have fiddled with an <a href="https://github.com/michalczaplinski/gutenberg-server-components">experiment that allows you to create server components in PHP with Gutenberg </a> and I feel like it's not that promising. Perhaps a better way would be to explore something like <a href="https://laravel-livewire.com/"> Laravel Livewire </a>.</li><li>If we were to go this route, what direction would it make sense to explore with the new block format? Should we try to create our own format or adopt a format that already exists (Mitosis, Svelte, or something else)?</li></ul>

**Comment by michalczaplinski** [comment]
I'd like to note another approach I have considered but ultimately decided against, because I've stumbled across a [PR to Jetpack](https://github.com/Automattic/jetpack/pull/13070) where it was already tried 🙂. 

The idea is to prerender React components to html templates at build time using React's <code>renderToStaticMarkup</code>. This approach works because at _build_ time, the user has access to nodejs (JS on the server) and we can execute <code>renderToStaticMarkup(block)</code> where <code>block</code> is our <strong>shared </strong>block. You also have to pass some placeholders for your props that can later be "filled out" at request time on the server.

```js
// This runs at build time. In the actual code we wouldn't save the 
// file manually but rather let webpack generate it for us. 

const markup = renderToStaticMarkup(
    <UpgradeNudge planName="#planName#" upgradeUrl="#upgradeUrl#" />
);

saveTemplate(markup, "_inc/blocks/upgrade.html");
```

```php
// This runs on the server at request time.
// The actual code is a little more complicated to also handle i18n and some edge cases.

ob_start();
include JETPACK__PLUGIN_DIR . "_inc/blocks/upgrade.html";
$markup = ob_get_clean();
foreach ( $props as $key => $value ) {
	$markup = str_replace(
		"#$key#",   // <---- like `planName` in the code snippet above
		$value,
		$markup
	);
}

echo $markup;
```

<p>However, this approach would only work for very simple blocks where the props do not determine the structure of the final HTML. If you have a component like this:

```jsx
const DogsList = ({ dogs }) => (
    <ul>
        {dogs.map(({ name }) => (
            <li>
                {name} {name === "Max" ? "is NOT a good dog." : "is a good dog!"}
            </li>
        ))}
    </ul>
);
```

this approach would not work because you cannot know before request-time how many dogs there are and if <code>Max</code> is one of them 🙂.

In order to be able to generate a template that could handle such components we would need a full-blown compiler that could turn the <code>{ dogs.map(({ name }) =&gt; component }</code> into an equivalent structure in PHP or into a template (like Mitosis does).

**Comment by michalczaplinski** [comment]
The proposal in OP (and the related demo) is one specific way of attacking the problem of block hydration. 

I'd like to step back and note some more abstract steps or "building blocks" that I think going to be necessary for a full solution. This is a non-exhaustive list, of course 🙂 


1. **Use the same template on the client and server.**
Whatever this template is, be it JSX or PHP or Liquid (like in the OP demo) It doesn't even have to be the "same" as in "identical", but rather "compatible".

2. **Identify the** _dynamic_ **and** _static_ **parts of the template.**
There are many approaches here. We use compilation in the OP, but others have used some type of light marking of the dynamic "holes" in the static templates (for example this was the approach [tried in Jetpack](https://github.com/Automattic/jetpack/pull/13070) that I've mentioned earlier)

3. **Enable looping and conditionals in the template.**
Mitosis has the [control flow components](https://github.com/BuilderIO/mitosis/blob/main/docs/components.md#control-flow) for that purpose. Each templating language has a flavour of those constructs and with JSX you use regular JS constructs to achieve this in userland.

4. **Make the** _dynamic_ **parts interactive on the client**
This is the actual "hydration" part of "block hydration" :)

**Comment by johnhooks** [comment]
@michalczaplinski to your knowledge have any of these projects defined a block AST?

**Comment by johnhooks** [comment]
This is my summary of the problem to see if I understand it correctly.

The power of JSX is that it is just JavaScript after transpiling. Though, this makes using it as a templating language for PHP very problematic because it practically requires transpiling JS to PHP.

Creating a new templating language or using an existing one would complicate the ecosystem of React based blocks.

The common target is serialized blocks.

Server rendering is required to generate markup for good SEO and quickly displaying properly formatted content on initial page load. The interactively has to be able to hydrate from this markup.

The solution needs to render in PHP, hydrate in JS, and remain compatible with the React block ecosystem.

**Reply by luisherranz** [reply]
> The solution needs to render in PHP, hydrate in JS, and remain compatible with the React block ecosystem.

Good summary 🙂

**Reply by johnhooks** [reply]
Thanks for the confirmation I'm thinking of this correctly,

**Comment by johnhooks** [comment]
My thoughts on the problem.

## The concept

- Create a JSX to PHP compiler, half of the job is already done. Robust AST can be generated from JSX.
- Start with a subset of JSX functionality.
- Use ESLint rules to help authors write complaint JSX.
- Start small, but eventually the functionality could become quite robust.

## The process

1. Generate PHP templates from JSX that accept the block attributes as props
2. Assign as the `render_callback` of a block type.
3. Hydrate islands of blocks in the DOM, using the same function that generated it.

## Pros

- Write views in JSX, the same way it is currently done.
- Define frontend interactivity in React.
- Retain the entire ecosystem of React and JS.

## Cons

- Complex - There will be a mountain of edge cases.
- Confusion - The view would appear to be standard JSX, a developer unaware of the difference could easily mistake it for regular JSX and become frustrated. (This could be partially circumnavigate by helpful hints from the compiler and `wp-scripts`)
- Third Party React Libraries - One of the awesome powers of React is the expansive functionality provided by third party code. Ensuring it can be appropriately compiled will most likely require a complete JS to PHP compilation.

## The questions this raises

Could the `save` function of a block type also be used to defined the client side interactivity, as well as, continue to fulfill its role as the save function in the editor?

Is loading React on the frontend an option?

## Revision

> However, there's a fundamental issue: the HTML saved in the database gets modified during WordPress rendering. - @luisherranz 

The `save` function cannot be used to both save the serialized block to the database and rehydrate the block on the frontend of a site, the markup is past through filters on the server that can modify the final output. The changes applied by the filters would be removed as soon as the block rehydrates.

**Reply by johnhooks** [reply]
One answer to third party client only code is to do like Astro and flag it as a `client-only` dependency to import in the client.

**Reply by luisherranz** [reply]
The first thing we tried was precisely using the block's save function for both defining client-side interactivity and the save in the editor ([custom elements hydration](https://github.com/WordPress/block-interactivity-experiments/issues/39)). However, there's a fundamental issue: the HTML saved in the database gets modified during WordPress rendering.

Consider this hypothetical component, used both to save the HTML and later for hydration of an interactive block:

```js
const MyBlock = () => {
  const [open, setOpen] = useState(true);

  return (
    <div>
      <button onClick={() => setOpen(!open)}>Toggle</button>
      {open && <span>Hi there!</span>}
    </div>
  );
};
```

And imagine a PHP filter that adds the class `"dynamic"` to the `<span>`:

```php
add_filter( 'render_block_my_block', function( $content ) {
  $p = new WP_HTML_Tag_Processor( $content );
  $p->next_tag( array( 'tag_name' => 'SPAN' ) );
  $p->add_class( 'dynamic' );
  return $p->get_updated_html();
} );
```

The HTML sent to the client will have the `"dynamic"` class, which is correct. But once the block is hydrated and the user clicks the toggle, the class disappears because it doesn't exist in the JSX.

So, a template DSL that compiles to JS templates isn't compatible with PHP filters, and WordPress heavily relies on them.

That's one of the main reasons we started investigating [directives](https://github.com/WordPress/block-interactivity-experiments/issues/80), where the templating language isn't JSX (or another template DSL) but the actual HTML itself.

**Reply by johnhooks** [reply]
Actually after thinking on it further, there are a lot of similarities between the requirements of this system and how Astro works. It uses a modified jsx syntax for `.astro` files. It take advantage of all the jsx tooling, it does require a special extension in VSCode. 

Using the same method would help with the con of developer confusion. If the templating language looks and feels like jsx, the developer would feel comfortable with it, but will be aware it is not the same thing because it uses a different file extension.

**Reply by johnhooks** [reply]
@luisherranz thank you for that detailed explanation, I see that definitely complicates the situation.

**Reply by johnhooks** [reply]
@luisherranz I do really like the idea of directives, but it's hard to accept a system of defining the edit function in React and the save function in PHP and directives. I'm sure you are aware and is one of the cons in your own list. Also, keep in mind I'm totally fresh to this problem, I'm not trying to state the obvious to you.

**Reply by gziolo** [reply]
It’s a long process to get to the templating language while supporting all existing blocks. That’s definitely our goal!

**Reply by luisherranz** [reply]
It's the main con, yes (currently) 🙂

**Reply by johnhooks** [reply]
@gziolo oh yeah, I understand this is not going to be a quick fix. But I'm interesting and plan to try to help with this problem. @luisherranz I said it would be "hard" to accept but I might be leaning that direction, the wrinkle of filters is something I didn't consider. I'm still getting used to how "pluggable" WordPress is!

**Reply by johnhooks** [reply]
A templating system that compiles to React and PHP and uses directives for all interactivity. Do you think this could be possible? To write blocks in some other templating language and still use editor components and connect to the Redux store in the edit function? (Sorry, I'm going to keep framing this in a way to have a unified view authoring experience)

**Reply by gziolo** [reply]
We are currently exploring whether it would make sense to bring the concept of directives also to JS by leveraging it in `edit` and `save` in https://github.com/WordPress/gutenberg/pull/48590. In general, we need to build lower-level primitives we can use as compilation targets, as this will inform the direction for the templating system people know from using Astro, Svelte, Solid, or Vue.

**Reply by johnhooks** [reply]
Thank you for helping me find the relevant Issues in the repo, it really helps. It's been hard to find a comprehensive explanation of the problem and the current solutions being considered.

**Reply by johnhooks** [reply]
I've only been in the WordPress space for two months. But the first thing I wanted to do after creating my first block was to add interactivity to the frontend. It has take a while to grasp why doing that is difficult, learn it is an acknowledged problem, and that work is being done to find a solution.

**Comment by johnhooks** [comment]
Correct me if I'm wrong. But I haven't seen a detailed Make blog post about this issue. Or detailed explanations of the proposed solutions. Do you think there is a more discoverable way to post this, rather than it being mostly contained in GitHub Issues and Discussions?

CC: @gziolo @michalczaplinski @luisherranz

**Reply by johnhooks** [reply]
If I wrote something up, would you be able to review it and help find the best location to publish it?

**Reply by johnhooks** [reply]
I know this post exists, but it's pretty vague on the details.

https://make.wordpress.org/core/2022/04/27/exploration-to-enable-better-developer-and-visitor-experiences-with-blocks/

**Reply by gziolo** [reply]
Great timing! You can expect a very detailed update on Make Core blog this week that will cover many concepts that are essential for universal blocks.

**Reply by johnhooks** [reply]
I'm excited to read it!

**Reply by michalczaplinski** [reply]
Here's the blog post 🙂  https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/

**Reply by johnhooks** [reply]
Thank you! That was great, and I really appreciate everyone taking the time to walk me through it, even with the big announcement post just about to go out.

**Comment by johnhooks** [comment]
> The solution needs to render in PHP, hydrate in JS, and remain compatible with the React block ecosystem.

**Revision**

The solution needs to render in PHP, _remain modifiable by WordPress filters on the server_, hydrate in JS, and remain compatible with the React block ecosystem.

**Comment by johnhooks** [comment]
Functional views are not a good fit for the use case of content that can be modified by a filter. React is the opposite concept. When the same things go in, same thing comes out. The system is broken if the output of the function can be modified. The same inputs no longer provide the same output.

> So, a template DSL that compiles to JS templates isn't compatible with PHP filters, and WordPress heavily relies on them.
> @luisherranz 

I missed that on the first read through. I think I get it now.

**Comment by michalczaplinski** [comment]
It looks like people keep (re)discovering this discussion, so I'd like to clarify how my thinking about Universal blocks has evolved over the last two years:

A group of contributors has been working on the [Interactivity API](https://make.wordpress.org/core/2024/03/04/interactivity-api-dev-note/) over the last year which is available in Wordpress 6.5. This API is a **direct result** of our thinking about how to make it easier to create interactive blocks (Challange 2. in the OP)

This API is available today, is compatible with all PHP filters, and works great with server-rendered content, unlike all of the React solutions suggested in this discussion.

That said, there is still a possible path to a universal block compiler in the long term. Such a compiler could, in principle, take as input something like `UI state` + `template` and return both the frontend and backend representations. Those could roughly be:
1. An interactive JS [store](https://github.com/WordPress/gutenberg/blob/trunk/packages/interactivity/docs/api-reference.md#the-store) using Interactivity API.
2. The server definition of the store ([link](https://github.com/WordPress/gutenberg/blob/trunk/packages/interactivity/docs/api-reference.md#on-the-server-side)).
3. PHP code that renders the HTML with [directives](https://github.com/WordPress/gutenberg/blob/trunk/packages/interactivity/docs/api-reference.md#list-of-directives).

However, no team is working on this at the moment and it might never happen. It largely depends on the adoption of the Interactivity API and the community feedback.

**Reply by ts-mz** [reply]
As per your last sentence - I wonder if having this kind (even development thereof) of universal block compiler would not also help increase the adoption of the interactivity API?

**Reply by michalczaplinski** [reply]
Hard to say 🙂
